src/main.rs

```diff
@@ -57,124 +49,64 @@
-    Exec(Exec),
+// This defines the toplevel `enarx` CLI
+#[derive(StructOpt, Debug)]
+#[structopt(
+    setting = structopt::clap::AppSettings::DeriveDisplayOrder,
+)]
+struct Options {
+    /// Logging options
+    #[structopt(flatten)]
+    log: cli::LogOptions,
+
+    /// Subcommands (with their own options)
+    #[structopt(flatten)]
+    cmd: cli::Command,
 }
 
 #[allow(clippy::unnecessary_wraps)]
 fn main() -> Result<()> {
-    let backends: &[Box<dyn Backend>] = &[
-        #[cfg(feature = "backend-sgx")]
-        Box::new(backend::sgx::Backend),
-        #[cfg(feature = "backend-kvm")]
-        Box::new(backend::kvm::Backend),
-    ];
-
-    let workldrs: &[Box<dyn Workldr>] = &[
-        #[cfg(feature = "wasmldr")]
-        Box::new(workldr::wasmldr::Wasmldr),
-    ];
-
-    match Options::from_args() {
-        Options::Info(_) => info(backends),
-        Options::Exec(e) => {
-            // FUTURE: accept tenant-provided shim, or fall back to builtin..
-            let backend = backend(backends);
-            let shim_bytes = backend.shim();
-            if let Some(path) = e.code {
-                let map = mmarinus::Kind::Private.load::<mmarinus::perms::Read, _>(&path)?;
-                exec(backend, shim_bytes, map)
-            } else {
-                exec(backend, shim_bytes, workldr(workldrs).exec())
-            }
+    let opts = Options::from_args();
+    opts.log.init_logger();
+
+    info!("logging initialized!");
+    info!("CLI opts: {:?}", &opts);
+
+    match opts.cmd {
+        cli::Command::Info(info) => info.display(),
+        cli::Command::Exec(exec) => {
+            let backend = exec.backend.pick()?;
+            let binary = mmarinus::Kind::Private.load::<mmarinus::perms::Read, _>(&exec.binpath)?;
+            keep_exec(backend, backend.shim(), binary)
         }
-    }
-}
-
-#[allow(clippy::unnecessary_wraps)]
-fn info(backends: &[Box<dyn Backend>]) -> Result<()> {
-    use colorful::*;
-
-    for backend in backends {
-        println!("Backend: {}", backend.name());
-
-        let data = backend.data();
-
-        for datum in &data {
-            let icon = match datum.pass {
-                true => "✔".green(),
-                false => "✗".red(),
-            };
-
-            if let Some(info) = datum.info.as_ref() {
-                println!(" {} {}: {}", icon, datum.name, info);
-            } else {
-                println!(" {} {}", icon, datum.name);
-            }
-        }
-
-        for datum in &data {
-            if let Some(mesg) = datum.mesg.as_ref() {
-                println!("\n{}\n", mesg);
-            }
+        cli::Command::Run(run) => {
+            let modfile = File::open(run.module)?;
+            let open_fd = modfile.as_raw_fd();
+            // TODO: we should pass the fd to workldr, but we don't actually
+            // have a way to pass arguments to workldr yet, so for now..
+            assert!(open_fd == 3, "module got unexpected fd {}", open_fd);
```

Comment:
Shouldn't this result in an error, rather than a panic? (the outer function returns a `Result`)